internal/evmreader/evmreader.go

```diff
@@ -0,0 +1,344 @@
+	Application        = model.Application
+	Input              = model.Input
+	DefaultBlock       = model.DefaultBlock
+	InputBoxInputAdded = inputbox.InputBoxInputAdded
+	FilterOpts         = bind.FilterOpts
+	Context            = context.Context
+	Header             = types.Header
+	Subscription       = ethereum.Subscription
+)
+
+// Interface for Input reading
+type InputSource interface {
+	// Wrapper for FilterInputAdded(), which is automatically generated by go-ethereum and cannot be used for testing
+	RetrieveInputs(
+		opts *FilterOpts,
+		appContract []Address,
+		index []*big.Int,
+	) ([]InputBoxInputAdded, error)
+}
+
+// Interface for the node repository
+type EvmReaderRepository interface {
+	InsertInputsAndUpdateLastProcessedBlock(
+		ctx Context,
+		inputs []Input,
+		blockNumber uint64,
+		appAddress Address,
+	) error
+	GetAllRunningApplications(
+		ctx context.Context,
+	) ([]Application, error)
+}
+
+// EthClient mimics part of ethclient.Client functions to narrow down the
+// interface needed by the EvmReader. It must be bound to an HTTP endpoint
+type EthClient interface {
+	HeaderByNumber(
+		ctx Context,
+		number *big.Int,
+	) (*Header, error)
+}
+
+// EthWsClient mimics part of ethclient.Client functions to narrow down the
+// interface needed by the EvmReader. It must be bound to a WS endpoint
+type EthWsClient interface {
+	SubscribeNewHead(
+		ctx Context,
+		ch chan<- *Header,
+	) (Subscription, error)
+}
+
+type SubscriptionError struct {
+	Cause error
+}
+
+func (e *SubscriptionError) Error() string {
+	return fmt.Sprintf("Subscription error : %v", e.Cause)
+}
+
+type nodeConfig struct {
+	inputBoxBlockNumber uint64
+	defaultBlock        DefaultBlock
+}
+
+// EvmReader reads inputs from the blockchain
+type EvmReader struct {
+	client      EthClient
+	wsClient    EthWsClient
+	inputSource InputSource
+	repository  EvmReaderRepository
+	config      nodeConfig
+}
+
+func (r *EvmReader) String() string {
+	return "evm-reader"
+}
+
+// Creates a new EvmReader
+func NewEvmReader(
+	client EthClient,
+	wsClient EthWsClient,
+	inputSource InputSource,
+	repository EvmReaderRepository,
+	inputboxBlockNumber uint64,
+	defaultBlock DefaultBlock,
+
+) EvmReader {
+	return EvmReader{
+		client:      client,
+		wsClient:    wsClient,
+		inputSource: inputSource,
+		repository:  repository,
+		config: nodeConfig{
+			inputBoxBlockNumber: inputboxBlockNumber,
+			defaultBlock:        defaultBlock,
+		},
+	}
+}
+
+func (r *EvmReader) Start(
+	ctx context.Context,
+	ready chan<- struct{},
+) error {
+
+	for {
+		watchForNewInputsError := r.watchForNewInputs(ctx, ready)
+		if _, ok := watchForNewInputsError.(*SubscriptionError); !ok {
+			return watchForNewInputsError
+		}
+		slog.Debug(watchForNewInputsError.Error())
+		slog.Debug("Reconnecting...")
+	}
+}
+
+// Watch for new blocks and reads new inputs based on the default block configuration, which have not
+// been processed yet.
+func (r *EvmReader) watchForNewInputs(
```

Comment:
I find this name confusing, it's too similar to `checkForNewInputs`. Either this one should be named something more like `subscribeForNewInputs`, or `checkForNewInputs` should be named something like `processNewInputs`